cmd/pyroscope/command/common.go

```diff
@@ -0,0 +1,32 @@
+package command
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/pyroscope-io/pyroscope/pkg/cli"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+func loadConfig(path string, v interface{}) error {
+	if path == "" {
+		return nil
+	}
+
+	viper.SetConfigFile(path)
+	if err := viper.ReadInConfig(); err != nil {
+		return err
+	}
+
+	return viper.Unmarshal(v)
+}
+
+func loadFlags(cfg interface{}, cmd *cobra.Command, opts ...cli.FlagOption) {
+	cli.PopulateFlagSet(cfg, cmd.Flags(), opts...)
+	viper.BindPFlags(cmd.Flags())
+
+	if err := viper.Unmarshal(cfg); err != nil {
+		fmt.Fprintln(os.Stderr, "Unable to unmarshal:", err)
+	}
+}
```

Comment:
I'm afraid but use of global `viper` instance is not we really want. In `PopulateFlagSet` we set default values by parameter name, like:
```golang
	val := fieldV.Addr().Interface().(*string)
	for old, n := range o.replacements {
		defaultValStr = strings.ReplaceAll(defaultValStr, old, n)
	}
	flagSet.StringVar(val, nameVal, defaultValStr, descVal)
	viper.SetDefault(nameVal, defaultValStr)
```

The problem is that it happens during a command initialisation (when `loadFlags` is called) so that _all_ flag sets for _all_ commands are processed regardless of the subcommand user invokes. And due to name collisions (e.g. `config`), everything goes really bad. For example, `pyroscope agent` configuration file is resolved to `C:\Program Files\Pyroscope/etc/pyroscope/server.yml`.

The problem can be solved by using separate `viper` instances for every subcommand.